x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/huggingface/response/HuggingFaceRerankResponseEntity.java

```diff
@@ -0,0 +1,130 @@
+     *                   }
+     *               ]
+     *          }
+     *     </code>
+     * </pre>
+     */
+
+    public static RankedDocsResults fromResponse(HuggingFaceRerankRequest request, HttpResult response) throws IOException {
+        var parserConfig = XContentParserConfiguration.EMPTY.withDeprecationHandler(LoggingDeprecationHandler.INSTANCE);
+
+        try (XContentParser jsonParser = XContentFactory.xContent(XContentType.JSON).createParser(parserConfig, response.body())) {
+            moveToFirstToken(jsonParser);
+
+            XContentParser.Token token = jsonParser.currentToken();
+            ensureExpectedToken(XContentParser.Token.START_ARRAY, token, jsonParser);
+
+            var rankedDocs = doParse(jsonParser);
+            var rankedDocsByRelevanceStream = rankedDocs.stream()
+                .sorted(Comparator.comparingDouble(RankedDocsResults.RankedDoc::relevanceScore).reversed());
+            var rankedDocStreamTopN = request.getTopN() == null
+                ? rankedDocsByRelevanceStream
+                : rankedDocsByRelevanceStream.limit(request.getTopN());
+            return new RankedDocsResults(rankedDocStreamTopN.toList());
+        }
+    }
+
+    private static List<RankedDocsResults.RankedDoc> doParse(XContentParser parser) throws IOException {
+        return parseList(parser, (listParser, index) -> {
+            var parsedRankedDoc = HuggingFaceRerankResponseEntity.RankedDocEntry.parse(parser);
+
+            if (parsedRankedDoc.id == null) {
+                throw new IllegalStateException(format(FAILED_TO_FIND_FIELD_TEMPLATE, RankedDocEntry.ID.getPreferredName()));
+            }
+
+            if (parsedRankedDoc.score == null) {
+                throw new IllegalStateException(format(FAILED_TO_FIND_FIELD_TEMPLATE, RankedDocEntry.SCORE.getPreferredName()));
+            }
+
+            try {
+                return new RankedDocsResults.RankedDoc(parsedRankedDoc.id, parsedRankedDoc.score, parsedRankedDoc.text);
+            } catch (NumberFormatException e) {
+                throw new IllegalStateException(format(INVALID_ID_FIELD_FORMAT_TEMPLATE, parsedRankedDoc.id));
+            }
+        });
+    }
+
+    private record RankedDocEntry(@Nullable Integer id, @Nullable Float score, @Nullable String text) {
+
+        private static final ParseField TEXT = new ParseField("text");
+        private static final ParseField SCORE = new ParseField("score");
+        private static final ParseField ID = new ParseField("index");
+        private static final ConstructingObjectParser<HuggingFaceRerankResponseEntity.RankedDocEntry, Void> PARSER =
+            new ConstructingObjectParser<>(
+                "hugging_face_rerank_response",
+                true,
+                args -> new HuggingFaceRerankResponseEntity.RankedDocEntry((int) args[0], (float) args[1], (String) args[2])
+            );
+
+        static {
+            PARSER.declareInt(ConstructingObjectParser.constructorArg(), ID);
```

Comment:
`constructorArg()` means that the argument is required. Let's remove the `@Nullable` for `id` and `score`.